File: app/src/main/java/com/example/thekhaeng/viewstatesavetest/view/SwitchViewGroup.java
package com.example.thekhaeng.viewstatesavetest.view;

import android.content.Context;
import android.os.Build;
import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.RequiresApi;
import android.text.TextWatcher;
import android.util.AttributeSet;
import android.widget.CompoundButton;
import android.widget.EditText;

import com.example.thekhaeng.viewstatesavetest.R;
import com.example.thekhaeng.viewstatesavetest.testSavedStateAdapter.adapter.UserAdapter;
import com.example.thekhaeng.viewstatesavetest.view.base.BaseViewGroup;

/**
 * Created by TheKhaeng
 */

public class SwitchViewGroup extends BaseViewGroup{

    private EditText editText;
    private CustomSwitch toggle;
    private UserAdapter.OnCheckedChangeListener checkChangeListener;
    private UserAdapter.OnTextChangeListener textChangedListener;

    public SwitchViewGroup( Context context, AttributeSet attrs ){
        super( context, attrs );
    }

    public SwitchViewGroup( Context context, AttributeSet attrs, int defStyleAttr ){
        super( context, attrs, defStyleAttr );
    }

    @RequiresApi( api = Build.VERSION_CODES.LOLLIPOP )
    public SwitchViewGroup( Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes ){
        super( context, attrs, defStyleAttr, defStyleRes );
    }

    @Override
    protected int getLayoutRes(){
        return R.layout.switch_custom_view_group;
    }

    @Override
    protected void setupStyleables( AttributeSet attrs, int defStyleAttr, int defStyleRes ){
    }

    @Override
    protected void bindView(){
        editText = (EditText) findViewById( R.id.text );
        toggle = (CustomSwitch) findViewById( R.id.toggle );
    }

    @Override
    protected void setupView(){
    }

    public void setText( String message ){
        editText.setText( message );
    }

    public void setTextChangedListener( TextWatcher listener ){
        editText.addTextChangedListener( listener );
    }

    public void setOnCheckChangeListener( CompoundButton.OnCheckedChangeListener listener ){
        toggle.setOnCheckedChangeListener( listener );
    }

    public void setCheck( boolean check ){
        toggle.setChecked( check );
    }


    @Override
    public Parcelable onSaveInstanceState(){
        Parcelable superState = super.onSaveInstanceState();
//        SavedState ss = new SavedState( superState );
        // Must call
        SavedState ss = (SavedState) onSaveInstanceChildState( new SavedState( superState ) );
        //no data to save
        return ss;
    }


    @Override
    public void onRestoreInstanceState( Parcelable state ){
        if( !( state instanceof SavedState ) ){
            super.onRestoreInstanceState( state );
            return;
        }
        SavedState ss = (SavedState) state;
        super.onRestoreInstanceState( ss );
        //no data to restore
    }


    private static class SavedState extends ChildSavedState{

        SavedState( Parcelable superState ){
            super( superState );
        }

        private SavedState( Parcel in, ClassLoader classLoader ){
//            super( in );
            super( in, classLoader );
            //no data to save
        }

        @Override
        public void writeToParcel( Parcel out, int flags ){
            super.writeToParcel( out, flags );
            //no data to restore
        }

        public static final ClassLoaderCreator<SavedState> CREATOR = new ClassLoaderCreator<SwitchViewGroup.SavedState>(){
            @Override
            public SwitchViewGroup.SavedState createFromParcel( Parcel source, ClassLoader loader ){
                return new SwitchViewGroup.SavedState( source, loader );
            }

            public SwitchViewGroup.SavedState createFromParcel( Parcel in ){
                return null;
            }


            public SwitchViewGroup.SavedState[] newArray( int size ){
                return new SwitchViewGroup.SavedState[size];
            }
        };
    }

}

File: app/src/main/java/com/example/thekhaeng/viewstatesavetest/view/kotlin/SwitchViewGroup.kt
package com.example.thekhaeng.viewstatesavetest.view.kotlin

import android.content.Context
import android.os.Parcel
import android.os.Parcelable
import android.text.TextWatcher
import android.util.AttributeSet
import android.widget.CompoundButton
import android.widget.EditText
import com.example.thekhaeng.viewstatesavetest.R
import com.example.thekhaeng.viewstatesavetest.testSavedStateAdapter.adapter.UserAdapter
import com.example.thekhaeng.viewstatesavetest.view.kotlin.base.BaseViewGroup

/**
 * Created by TheKhaeng
 */

class SwitchViewGroup
@JvmOverloads constructor(context: Context,
                          attrs: AttributeSet? = null,
                          defStyleAttr: Int = 0,
                          defStyleRes: Int = 0)
    : BaseViewGroup(context, attrs, defStyleAttr, defStyleRes) {

    private var editText: EditText? = null
    private var toggle: CustomSwitch? = null
    private val checkChangeListener: UserAdapter.OnCheckedChangeListener? = null
    private val textChangedListener: UserAdapter.OnTextChangeListener? = null


    override
    fun getLayoutRes(): Int = R.layout.switch_custom_view_group

    override
    fun setupStyleables(attrs: AttributeSet, defStyleAttr: Int, defStyleRes: Int) {}

    override
    fun bindView() {
        editText = findViewById(R.id.editText)
        toggle = findViewById(R.id.toggle)
    }

    override
    fun setupView() {}

    fun setTextToEditText(message: String?) {
        editText?.setText(message)
    }

    fun setTextChangedListener(listener: TextWatcher) {
        editText?.addTextChangedListener(listener)
    }

    fun setOnCheckChangeListener(listener: CompoundButton.OnCheckedChangeListener) {
        toggle?.setOnCheckedChangeListener(listener)
    }

    fun setCheck(check: Boolean) {
        toggle?.isChecked = check
    }

    public override fun onSaveInstanceState(): Parcelable? {
        val superState = super.onSaveInstanceState()
        // Must call
        val ss = onSaveChildInstanceState(SavedState(superState)) as SavedState
        // save data here
        return ss
    }


    public override fun onRestoreInstanceState(state: Parcelable) {
        if (state !is SavedState) {
            super.onRestoreInstanceState(state)
            return
        }
        super.onRestoreInstanceState(state)
        // restore data here
    }


    private class SavedState : BaseViewGroup.ChildSavedState {

        internal constructor(superState: Parcelable) : super(superState) {}

        private constructor(`in`: Parcel, classLoader: ClassLoader) : super(`in`, classLoader) {
            // save data here
        }

        override
        fun writeToParcel(out: Parcel, flags: Int) {
            super.writeToParcel(out, flags)
            // restore data here
        }

        companion object {
            @JvmField
            val CREATOR: Parcelable.ClassLoaderCreator<SavedState> = object : Parcelable.ClassLoaderCreator<SavedState> {
                override
                fun createFromParcel(source: Parcel, loader: ClassLoader): SavedState = SavedState(source, loader)

                override
                fun createFromParcel(`in`: Parcel): SavedState? = null

                override
                fun newArray(size: Int): Array<SavedState?> = arrayOfNulls(size)
            }
        }
    }

}
